include/verilated_imp.h

```diff
@@ -448,36 +447,105 @@ class VerilatedImp {
+        FILE* fp;
         VerilatedLockGuard lock(s_s.m_fdMutex);
-        if (VL_UNLIKELY(!(fdi & (VL_ULL(1) << 31)) || idx >= s_s.m_fdps.size())) return;
-        if (VL_UNLIKELY(!s_s.m_fdps[idx])) return;  // Already free
-        s_s.m_fdps[idx] = NULL;
-        s_s.m_fdFree.push_back(idx);
-    }
-    static inline FILE* fdToFp(IData fdi) VL_MT_SAFE {
-        IData idx = VL_MASK_I(31) & fdi;
-        VerilatedLockGuard lock(s_s.m_fdMutex);  // This might get slow, if it does we can cache it
-        if (VL_UNLIKELY(!(fdi & (VL_ULL(1) << 31)) || idx >= s_s.m_fdps.size())) return NULL;
-        return s_s.m_fdps[idx];
+        const int n = fdToFp(fdi, &fp);
+        if(VL_UNLIKELY(!fp || (n != 1))) return 0;
+        return static_cast<IData>(fseek(fp, static_cast<long>(offset), static_cast<int>(origin)));
+    }
+    static IData fdTell(IData fdi) VL_MT_SAFE {
+        FILE* fp;
+        VerilatedLockGuard lock(s_s.m_fdMutex);
+        const int n = fdToFp(fdi, &fp);
+        if(VL_UNLIKELY(!fp || (n != 1))) return 0;
+        return static_cast<IData>(ftell(fp));
+    }
+    static void fdClose(IData fdi) VL_MT_SAFE {
+        VerilatedLockGuard lock(s_s.m_fdMutex);
+        if ((fdi & (1 << 31)) != 0) {
+            // Non-MCD case
+            IData idx = VL_MASK_I(31) & fdi;
+            if (VL_UNLIKELY(idx >= s_s.m_fdps.size())) return;
+            if (VL_UNLIKELY(!s_s.m_fdps[idx])) return;  // Already free
+            fclose(s_s.m_fdps[idx]);
+            s_s.m_fdps[idx] = nullptr;
+            s_s.m_fdFree.push_back(idx);
+        } else {
+            // MCD case
+            for (int i = 0; (fdi != 0) && (i < 31); i++, fdi >>= 1) {
+                if (fdi & VL_MASK_I(1)) {
+                    fclose(s_s.m_fdps[i]);
+                    s_s.m_fdps[i] = nullptr;
+                    s_s.m_fdFreeMct.push_back(i);
+                }
+            }
+        }
+    }
+    static inline int fdToFp(IData fdi, FILE** fp, std::size_t max = 1) VL_MT_SAFE {
+        if (VL_UNLIKELY(!fp || (max == 0))) return 0;
+        VerilatedLockGuard lock(s_s.m_fdMutex);
+        int out = 0;
+        if ((fdi & (1 << 31)) != 0) {
+            // Non-MCD case
+            IData idx = fdi & VL_MASK_I(31);
+            switch (idx) {
+            case 0: fp[out++] = stdin; break;
+            case 1: fp[out++] = stdout; break;
+            case 2: fp[out++] = stderr; break;
+            default:
+                if (VL_LIKELY(idx < s_s.m_fdps.size()))
+                    fp[out++] = s_s.m_fdps[idx];
+                break;
+            }
+        } else {
+            // MCD Case
+            for (int i = 0; (fdi != 0) && (out < max) && (i < 31); i++, fdi >>= 1)
+                if (fdi & VL_MASK_I(1)) fp[out++] = s_s.m_fdps[i];
+        }
+        return out;
+    }
+private:
+    static void initialize_fp_table() {
+        s_s.m_fdps.resize(31);
+        std::fill(s_s.m_fdps.begin(), s_s.m_fdps.end(), nullptr);
```

Comment:
No C++11 for another 6 months.

```suggestion
        std::fill(s_s.m_fdps.begin(), s_s.m_fdps.end(), NULL);
```